larpandoracontent/LArObjects/LArAdaBoostDecisionTree.h

```diff
@@ -0,0 +1,319 @@
+        const bool      m_outcome;              ///< Outcome if leaf node
+    };
+
+    typedef std::map<int, const Node*> IDToNodeMap;
+
+    class WeakClassifier
+    {
+    public: 
+        /**
+         *  @brief  Constructor, set hierarchy for nodes 
+         */
+        WeakClassifier(IDToNodeMap &idToNodeMap, const double &weight, const int &treeID);
+
+        /**
+         *  @brief  Predict signal or background based on trained data
+         *
+         *  @param  features the input features 
+         */
+        bool Predict(const DoubleVector &features) const;
+
+        /**
+         *  @brief  Evalute node and return outcome
+         *
+         *  @param  nodeID current node id 
+         *  @param  features the input features 
+         */
+        bool EvaluateNode(const int &nodeID, const DoubleVector &features) const;
+
+        /**
+         *  @brief  Get boost weight for weak classifier
+         */
+        double GetWeight() const;
+
+        /**
+         *  @brief  Get tree id for weak classifier
+         */
+        int GetTreeID() const;
+
+    private:
+        IDToNodeMap     m_idToNodeMap; ///< Decision tree nodes
+        const double    m_weight;      ///< Boost weight 
+        const int       m_treeID;      ///< Decision tree id        
+    };
+
+    typedef std::vector<const WeakClassifier*> WeakClassifiers;
+
+    class StrongClassifier
+    {
+    public:
+        /**
+         *  @brief  Constructor
+         */
+        StrongClassifier(WeakClassifiers &weakClassifiers);
+
+        /**
+         *  @brief  Predict signal or background based on trained data
+         *
+         *  @param  features the input features 
+         */
+        double Predict(const DoubleVector &features) const;
+
+    private: 
+        WeakClassifiers     m_weakClassifiers;     ///< Vector of weak classifers 
+    };
+
+    /**
+     *  @brief  Read the component at the current xml element
+     *
+     *  @param  pCurrentXmlElement address of the current xml element
+     *
+     *  @return success
+     */
+    pandora::StatusCode ReadComponent(pandora::TiXmlElement *pCurrentXmlElement);
+
+    /**
+     *  @brief  Read the decision tree component at the current xml handle
+     *
+     *  @param  currentHandle the current xml handle
+     *
+     *  @return success
+     */
+    pandora::StatusCode ReadDecisionTree(pandora::TiXmlHandle &currentHandle);
+
+    /**
+     *  @brief  Read the node component at the current xml handle
+     *
+     *  @param  currentHandle the current xml handle
+     *  @param  idToNodeMap map of id to node
+     *
+     *  @return success
+     */
+    pandora::StatusCode ReadNode(pandora::TiXmlHandle &currentHandle, IDToNodeMap &idToNodeMap);
+
+    WeakClassifiers       m_weakClassifiers;             ///< Vector of all weak classifiers
+    StrongClassifier     *m_pStrongClassifier;           ///< Strong adaptive boost tree classifier 
```

Comment:
Where/when are the memory allocations tidied up?

Reply:
Restructured code to ensure memory management is taken care of.